components/support/sync15-traits/src/bridged_engine.rs

```diff
@@ -87,134 +168,334 @@ pub trait BridgedEngine {
     // Don't provide access to the cleartext directly. We want all callers to
-    // use `IncomingEnvelope::payload`, so that we can validate the cleartext.
+    // use `IncomingEnvelope::payload`, so that we can validate the cleartext
+    // (where 'validate' in this context means 'let serde fail if it's invalid`)
     cleartext: String,
 }
 
-impl IncomingEnvelope {
-    /// Parses and returns the record payload from this envelope. Returns an
-    /// error if the envelope's cleartext isn't valid JSON, or the payload is
-    /// invalid.
-    pub fn payload(&self) -> Result<Payload, PayloadError> {
-        let payload: Payload = serde_json::from_str(&self.cleartext)?;
-        if payload.id != self.id {
-            return Err(PayloadError::MismatchedId {
-                envelope: self.id.clone(),
-                payload: payload.id,
-            });
+impl IncomingBridgeRecord {
+    /// Parses and returns an IncomingDeserPayload describing the record.
+    /// All JSON errors are treated as [IncomingDeserPayload::Malformed].
+    pub fn payload<T>(&self) -> IncomingDeserPayload<T>
+    where
+        T: DeserializeOwned,
+    {
+        let mut json = match serde_json::from_str(&self.cleartext) {
+            Ok(v) => v,
+            Err(_) => {
+                log::warn!("Invalid incoming json {}", self.envelope.id);
+                return IncomingDeserPayload::Malformed;
+            }
+        };
+
+        if let serde_json::Value::Object(ref map) = json {
+            if map.contains_key("deleted") {
+                return IncomingDeserPayload::Tombstone;
+            }
+        };
+
+        // In general, the payload does not carry 'id', but <T> does - so put
+        // it into the json before deserializing the record.
+        if let serde_json::Value::Object(ref mut map) = json {
+            match map.get("id").and_then(|v| v.as_str()) {
+                Some(id) => {
+                    // It exists in the payload! Note that this *should not* happen in practice
+                    // (the `id` should *never* be in the payload), but if that does happen
+                    // we should do the "right" thing, which is treat a mismatch as malformed.
+                    if id != self.envelope.id {
+                        log::warn!(
+                            "malformed incoming record: envelope id: {} payload id: {}",
+                            self.envelope.id,
+                            id
+                        );
+                        return IncomingDeserPayload::Malformed;
+                    }
+                    if !self.envelope.id.is_valid_for_sync_server() {
+                        log::warn!("malformed incoming record: id is not valid: {}", id);
+                        return IncomingDeserPayload::Malformed;
+                    }
+                    log::warn!("incoming record has 'id' in the payload - it does match, but is still unexpected");
+                }
+                None => {
```

Comment:
Note: This would capture the case where `id` exists as a non-string value, is it worth checking for that?

Reply:
yeah, I guess so, thanks.